pkg/suggestion/v1beta1/optuna/service.py

```diff
@@ -178,3 +184,37 @@ def _validate_random_setting(cls, algorithm_settings):
                                                                                        exception=e)
 
         return True, ""
+
+    @classmethod
+    def _validate_grid_setting(cls, experiment):
+        algorithm_settings = experiment.spec.algorithm.algorithm_settings
+        search_space = HyperParameterSearchSpace.convert(experiment)
+
+        for s in algorithm_settings:
+            try:
+                if s.name == "random_state":
+                    if not int(s.value) >= 0:
+                        return False, ""
+                else:
+                    return False, "unknown setting {} for algorithm grid".format(s.name)
+
+            except Exception as e:
+                return False, "failed to validate {name}({value}): {exception}".format(name=s.name, value=s.value,
+                                                                                       exception=e)
+
+        try:
+            combinations = HyperParameterSearchSpace.convert_to_combinations(search_space)
+            if isinstance(combinations, str):
+                return False, combinations
+
+            num_combinations = len(list(itertools.product(*combinations.values())))
+            max_trial_count = experiment.spec.max_trial_count
+            if max_trial_count > num_combinations:
+                return False, "Max Trial Count: {max_trial} > all possible search combinations: {combinations}".\
+                    format(max_trial=max_trial_count, combinations=num_combinations)
```

Comment:
I'm not familiar with optuna, Is it ok to set `max_trial_count < num_combinations` ? since with the documentation GridSampler looks like always trying to search all possible combinations.
- https://optuna.readthedocs.io/en/latest/reference/samplers/generated/optuna.samplers.GridSampler.html

Reply:
Yes, that's right about optuna behavior. Although optuna suggestion-service creates trials using already used parameters after all combinations used.

https://github.com/kubeflow/katib/blob/db72ce1987b11cac09d512587261187465ab2729/pkg/suggestion/v1beta1/optuna/base_service.py#L58-L65

However, if optuna suggestion services remove duplicated suggestions in the above, the suggestionclient faces the error in the following:

https://github.com/kubeflow/katib/blob/db72ce1987b11cac09d512587261187465ab2729/pkg/controller.v1beta1/suggestion/suggestionclient/suggestionclient.go#L123-L128

Also, if users want to run experiments using all combinations, users do not need to set `maxTrialCount` to the Experiment.

For these reasons, we would be good to provide that validation.

@anencore94 Does that make sense?

Reply:
Yeap I see now. Thanks for your kind reply :)